libcaf_core/caf/config_option.hpp

```diff
@@ -104,37 +104,13 @@ class CAF_CORE_EXPORT config_option {
   mutable void* value_;
 };
 
-/// Finds `config_option` string with a matching long name in (`first`, `last`],
-/// where each entry is a pointer to a string. Returns a `ForwardIterator` to
-/// the match and a `string_view` of the option value if the entry is
-/// found and a `ForwardIterator` to `last` with an empty `string_view`
-/// otherwise.
-template <class ForwardIterator, class Sentinel>
-std::pair<ForwardIterator, std::string_view>
-find_by_long_name(const config_option& x, ForwardIterator first,
-                  Sentinel last) {
-  auto long_name = x.long_name();
-  for (; first != last; ++first) {
-    std::string_view str{*first};
-    // Make sure this is a long option starting with "--".
-    if (!starts_with(str, "--"))
-      continue;
-    str.remove_prefix(2);
-    // Skip optional "caf#" prefix.
-    if (starts_with(str, "caf#"))
-      str.remove_prefix(4);
-    // Make sure we are dealing with the right key.
-    if (!starts_with(str, long_name))
-      continue;
-    // Make sure the key is followed by an assignment.
-    str.remove_prefix(long_name.size());
-    if (!starts_with(str, "="))
-      continue;
-    // Remove leading '=' and return the value.
-    str.remove_prefix(1);
-    return {first, str};
-  }
-  return {first, std::string_view{}};
-}
+/// Finds `config_option` string with a matching long name in the argument list
+/// [`first`, `last`). Returns an `iterator` to the matching location and a
+/// `string_view` of the option value if the entry is found, or a `iterator`
+/// to `last` with an empty `string_view` otherwise.
+std::pair<std::vector<std::string>::const_iterator, std::string_view>
+find_by_long_name(const config_option& x,
+                  std::vector<std::string>::const_iterator first,
+                  std::vector<std::string>::const_iterator last);
```

Comment:
Good call turning this into a non-template function. It would also need `CAF_CORE_EXPORT`, but I would actually go one step further and lift this into a member function. Something like this:

```cpp
  // ... inside config_option ...

  /// An iterator over CLI arguments.
  using argument_iterator = std::vector<std::string>::const_iterator;

  /// Stores the result of a find operation. The option sets `begin == end`
  /// if the operation could not find a match.
  struct find_result {
    /// The begin of the matched range.
    argument_iterator begin;
  
    /// The end of the matched range.
    argument_iterator end;

    /// The value for the config option.
    std::string_view value;
  };

  /// Tries to find this option by its long name in `[first, last)`.
  find_result find_by_long_name(argument_iterator first,
                                argument_iterator last) const noexcept;
```

This also addresses an issue I have in the current caller code. Currently, `actor_system_config` goes looking for '=' and then guesses that it has to erase `{i, i + 2}` from the vector. By returning the matched range instead, we can do:

```cpp
auto [first, last, path] = find_by_long_name(*ptr, args.begin(), args.end());
if (first == last) {
  // ... not found ...
} else {
  auto path_str = std::string{path};
  args.erase(first, last);
}
```

Reply:
I had some similar ideas, returning a std::pair<iterator, iterator>. What I'd like is refactoring this use-case completely away and use only the newer config_option_set::parse function, but that would require a lot more effort and less gain. Maybe that could be done if we ever get to #1440.
It's probably for the best if I go with the proposed changes.